File: src/lib/print.lua
include("kernel")
import("./../corelibs/terminal.js")
import("./print.js")

function print_int(val : int) : void --print integer
    if val < 0 then
        terminal_static_write("-")
        val = -val
    elseif val == 0 then
        terminal_static_write("0")
        return
    end

    local str = {}
    while val > 0 do
        str[#str + 1] = charCode("0") + val % 10
        val = val / 10
    end

    for i = #str, 1, -1 do
        terminal_write(str[i])
    end
end

function print_float(val : float) : void -- print float
    local base = js_eval("val.base")
    local left, right = base / 100, base % 100

    print_int(left)

    if right < 0 then
        right = -right
    end

    print_char(charCode("."))
    print_char(right / 10 + charCode("0"))
    print_char(right % 10 + charCode("0"))
end

--print char in print.js

function print_bool(val : bool) : void --print bool
    if val then
        terminal_static_write("true")
        return
    else
        terminal_static_write("false")
        return
    end
end

File: src/lib/print.js
var write = scope.get("terminal_write");
var writestr = scope.get("terminal_write_str");
var swrite = scope.get("terminal_static_write");
var swriteln = scope.get("terminal_static_writeln");

exports.print_char = function(c)
{
    write(c);
}

exports.print_str = function(s)
{
    writestr(s);
}

exports.printf = function(msg)
{
    var printFormats = {
        "s": exports.print_str,
        "f": scope.get("print_float"),
        "d": scope.get("print_int"),
        "i": scope.get("print_int"),
        "b": scope.get("print_bool"),
        "c": exports.print_char
    };

    var i = 1;
    var index = msg.indexOf("%");
    var start = 0;
    while(index != -1)
    {
        var format = msg[index + 1];
        if(printFormats.hasOwnProperty(format))
        {
            var before = msg.substring(0, index);
            if(before.length > 0)
                swrite(before);

            printFormats[format](arguments[i]);
            i++;
            msg = msg.substr(index + 2);
        }
        else
        {
            start += 2;
        }

        index = msg.indexOf("%", start);
    }
    if(msg.length > 0)
        swrite(msg);
}
